### PR TITLE
CORE-13796: Merge 5.0 into 5.1

### PR DESCRIPTION
**Pull request by josephzunigadaly**
Reminder: No squashing when integrating this PR

### CODE REVIEW

**Review by vkolomeyko — APPROVED** [review]
LGTM

**Review by vlajos — APPROVED** [review]
My change looks OK: components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/flow/impl/persistence/UtxoLedgerStateQueryServiceImpl.kt

**Review by kyriathar — APPROVED** [review]
https://github.com/corda/corda-runtime-os/pull/3994 ([c2b4eb1](https://github.com/corda/corda-runtime-os/pull/4035/commits/c2b4eb16a7df00c9c0f1f0950b4912b615839ccb)) looks good.

**Review by seanbrereton — APPROVED** [review]
LGTM (build)

**Review by davidcurrie — APPROVED** [review]
For `charts`

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt [hunk 1]
```diff
@@ -65,41 +69,62 @@ class EntityMessageProcessor(
...
+        val clientRequestId =
+            request.flowExternalEventContext.contextProperties.toMap()[MDC_CLIENT_ID] ?: ""
+
+        return withMDC(
+            mapOf(
+                MDC_CLIENT_ID to clientRequestId,
+                MDC_EXTERNAL_EVENT_ID to request.flowExternalEventContext.requestId
+            )
+        ) {
+            val startTime = Instant.now()
```
Comment (chrisr3) [inline components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt:98]: This is wrong - we should be using `System.nanoTime()` for calculating time elapsed.

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt [hunk 2]
```diff
@@ -65,41 +69,62 @@ class EntityMessageProcessor(
...
+
+                processRequestWithSandbox(sandbox, request).also { requestOutcome = "SUCCEEDED" }
+            } catch (e: Exception) {
+                responseFactory.errorResponse(request.flowExternalEventContext, e)
+            } finally {
+                CordaMetrics.Metric.Db.EntityPersistenceRequestTime.builder()
+                    .withTag(CordaMetrics.Tag.OperationName, request.request::class.java.name)
+                    .withTag(CordaMetrics.Tag.OperationStatus, requestOutcome)
+                    .build()
+                    .record(Duration.between(startTime, Instant.now()))
```
Comment (chrisr3) [inline components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt:121]: And this should be
```kotlin
.record(Duration.ofNanos(System.nanoTime() - startTime))
```

File: components/reconciliation/reconciliation-impl/src/main/kotlin/net/corda/reconciliation/impl/ReconcilerEventHandler.kt [hunk 3]
```diff
@@ -71,17 +73,34 @@ internal class ReconcilerEventHandler<K : Any, V : Any>(
 
     private fun reconcileAndScheduleNext(coordinator: LifecycleCoordinator) {
         logger.info("Initiating reconciliation")
+        var reconciliationOutcome = "FAILED"
+        val startTime = System.currentTimeMillis()
+        var reconciliationRunTime = startTime
```
Comment (chrisr3) [inline components/reconciliation/reconciliation-impl/src/main/kotlin/net/corda/reconciliation/impl/ReconcilerEventHandler.kt:78]: Again, we should not be calculating elapsed time using `System.currentTimeMillis()`.

File: components/reconciliation/reconciliation-impl/src/main/kotlin/net/corda/reconciliation/impl/ReconcilerEventHandler.kt [hunk 4]
```diff
@@ -71,17 +73,34 @@ internal class ReconcilerEventHandler<K : Any, V : Any>(
...
+            reconciledCount = reconcile()
+            reconciliationOutcome = "SUCCEEDED"
+            reconciliationRunTime = System.currentTimeMillis() - startTime
+            logger.info("Reconciliation completed in $reconciliationRunTime ms")
             scheduleNextReconciliation(coordinator)
-        } catch (e: Exception) {
+        } catch (e: Throwable) {
             // An error here could be a transient or not exception. We should transition to `DOWN` and wait
             // on subsequent `RegistrationStatusChangeEvent` to see if it is going to be a `DOWN` or an `ERROR`.
+            reconciliationRunTime = System.currentTimeMillis() - startTime
```
Comment (chrisr3) [inline components/reconciliation/reconciliation-impl/src/main/kotlin/net/corda/reconciliation/impl/ReconcilerEventHandler.kt:89]: This is also wrong.

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/StateAndEventConsumerImpl.kt [hunk 5]
```diff
@@ -261,7 +262,9 @@ internal class StateAndEventConsumerImpl<K : Any, S : Any, E : Any>(
     }
 
     override fun waitForFunctionToFinish(function: () -> Any, maxTimeout: Long, timeoutErrorMessage: String): CompletableFuture<Any> {
-        val future: CompletableFuture<Any> = CompletableFuture.supplyAsync({ function() }, executor)
+        val future: CompletableFuture<Any> = CompletableFuture.supplyAsync(
+            { function() },
```
Comment (chrisr3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/StateAndEventConsumerImpl.kt:266]: Surely just
```kotlin
function,
```
??

**Review by chrisr3 — COMMENTED** [review]
My changes look OK, but I thought I'd already updated our `Timer`s to use `System.nanoTime()` to calculate elapsed time where possible. Are these new occurrences of `System.currentTimeMillis()` from resolving a conflict?

File: components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt (hunk 1 above)
Comment (josephzunigadaly) [reply components/persistence/entity-processor-service-impl/src/main/kotlin/net/corda/entityprocessor/impl/internal/EntityMessageProcessor.kt:98]: These must be wrong in the source branch. I'll raise a PR to fix them then flag you for review. Once that is merged I'll pull it into this PR.

**Review by dimosr — APPROVED** [review]
approving on behalf of @corda/corda-platform-network-team to satisfy code owners rules, assuming authors will review their corresponding commits given it's a forward-merge.

**Review by lankydan — APPROVED** [review]

**Comment by josephzunigadaly** [comment]
I've heard from almost everyone. Going to merge this now. 

I've spoken to Chris and he is happy the concerns he raised will be fixed in the linked PR and appear in a follow up merge.

**Pull request merged**